otter/integration/lib/autoscale.py

```diff
@@ -430,119 +388,91 @@ def _get_the_ips(the_server_ids):
-        self, removed_ids, rcs, timeout=60, period=1, total_servers=None
-    ):
-        """Wait for the scaling group to reflect the true state of the tenant's
-        server states.  Out-of-band server deletions or servers which
-        transition to an ERROR state should eventually be removed from the
-        scaling group's list of active servers.
-
-        :param list removed_ids: A list of server IDs that we should expect
-            the scaling group to realize are gone eventually.
-        :param TestResources rcs: The test resources necessary to invoke API
-            calls and manage state.
-        :param int timeout: The number of seconds to wait before giving up.
-        :param int period: The number of seconds between each check for ID
-            removal.
-        :param int total_servers: If provided, the total number of servers in
-            your scaling group.  This parameter is used for error reporting
-            purposes only.
-        :return: It'll return the value of ``rcs`` if successful.  An exception
-            will be raised otherwise, including timeout.
+    @inlineCallbacks
+    def choose_random_servers(self, rcs, n):
         """
+        Choose n servers from the active servers on the scaling
+        group.
+        """
+        code, body = yield self.get_scaling_group_state(rcs, [200])
 
-        def check(state):
-            code, response = state
-            if code == 404:
-                raise BreakLoopException(
-                    "Scaling group appears to have disappeared"
-                )
-
-            active_ids = extract_active_ids(response)
-            for deleted_id in removed_ids:
-                if deleted_id in active_ids:
-                    raise TransientRetryError()
-
-            return rcs
-
-        def poll():
-            return self.get_scaling_group_state(rcs).addCallback(check)
-
-        if total_servers:
-            report = (
-                "Scaling group failed to reflect {} of {} servers removed."
-                .format(len(removed_ids), total_servers)
-            )
-        else:
-            report = (
-                "Scaling group failed to reflect {} servers removed."
-                .format(len(removed_ids))
-            )
+        ids = extract_active_ids(body)
+        returnValue(random.sample(ids, n))
 
-        return retry_and_timeout(
-            poll, timeout,
-            can_retry=terminal_errors_except(TransientRetryError),
-            next_interval=repeating_interval(period),
-            clock=reactor,
-            deferred_description=report,
+    def wait_for_state(self, rcs, matcher, timeout=600, period=10, clock=None):
+        """
+        Wait for the state on the scaling group to match the provided matchers,
+        specified by matcher.
+
+        :param rcs: a :class:`otter.integration.lib.resources.TestResources`
+            instance
+        :param matcher: A :mod:`testtool.matcher`, as specified in
```

Comment:
I do like that idea - the reason I'd hesitate though is that I like what comes out of `mismatch.describe()` - I would love to change retrier to take a callable instead of a straight description string, or otherwise report the last error.  The callables would each have to implement their own string for why the test failed, otherwise.

And I think @glyph wanted retry to optionally log every intermediate error that causes a retry, and the log messages might be useful for debugging why a test is failing.